solcore/optics/external_optics.py

```diff
@@ -4,21 +4,67 @@
 import types
 
 
+def calculate_absorption_rcwa(tmm_out):
+    all_z = tmm_out['position'] * 1e-9
+    all_abs = tmm_out['absorption'] / 1e-9
+
+    def diff_absorption(z):
+        idx = all_z.searchsorted(z)
+        idx = np.where(idx <= len(all_z) - 2, idx, len(all_z) - 2)
+        try:
+            z1 = all_z[idx]
+            z2 = all_z[idx + 1]
+
+            f = (z - z1) / (z2 - z1)
+
+            out = f * all_abs[:, idx] + (1 - f) * all_abs[:, idx + 1]
+
+        except IndexError:
+            out = all_abs[:, idx]
+
+        return out
+
+    all_absorbed = np.trapz(diff_absorption(all_z), all_z)
+
+    return diff_absorption, all_absorbed
+
+def absorbed(self, z):
+    out = self.diff_absorption(self.offset + z) * (z < self.width)
+    return out.T
+
 def solve_external_optics(solar_cell, options):
-    """ Calculates the reflection, transmission and absorption of a solar cell from external data, preparing the structure for further calculations. The external data that must be provided in the solar cell definition is:
+    """ Uses the reflection, transmission and absorption of a solar cell from
+    external data, preparing the structure for further calculations. The
+    external data that must be provided in the solar cell definition are:
 
-    external_reflected: a function that provides the fraction of reflected light at the specified wavelengths (in m)
-    external_absorbed: a function that provides an array with the differential absorption at a depth Z (in m) at the specified wavelengths.
+    external_reflected: a function that provides the fraction of reflected light 
+                        at the specified wavelengths (in m)
+    external_absorbed: a function that provides an array with the differential 
+                        absorption at a depth Z (in m) at the specified wavelengths.
 
-    Note that both are functions of a single variable - wavelength in the first case and position in the second - and that the latter provides as output an array at the wavelengths indicated in the options.
+    Note that both are functions of a single variable - wavelength in the first
+    case and position in the second - and that the latter provides as output an
+    array at the wavelengths indicated in the options.
+
+    Or either:
+
+    RAT: a dictionary containing  the R, A, T numpy arrays {'R':R, 'A':A', T':T}
```

Comment:
Is A here a function of wl and z or just wl, as R and T?

Reply:
A is a matrix, not even a function, with uniform spacing for the  espectral (wl) and position (z).